src/Npgsql/NpgsqlConnection.cs

```diff
@@ -719,21 +724,38 @@ internal Task Close(bool async, CancellationToken cancellationToken = default)
                 throw new ArgumentOutOfRangeException("Unknown connection state: " + FullState);
             }
 
-            // TODO: The following shouldn't exist - we need to flow down the regular path to close any
-            // open reader / COPY. See test CloseDuringRead with multiplexing.
-            if (Settings.Multiplexing && ConnectorBindingScope == ConnectorBindingScope.None)
+            
+            if (Settings.Multiplexing)
             {
-                // TODO: Consider falling through to the regular reset logic. This adds some unneeded conditions
-                // and assignment but actual perf impact should be negligible (measure).
-                Debug.Assert(Connector == null);
-                FullState = ConnectionState.Closed;
-                Log.Debug("Connection closed (multiplexing)");
-                OnStateChange(OpenToClosedEventArgs);
-                Volatile.Write(ref _closing, 0);
-                return Task.CompletedTask;
+                if (ConnectorBindingScope == ConnectorBindingScope.Transaction)
```

Comment:
The reason Close wasn't an async method before, and that it has a an async CloseAsync local method, is to avoid the async state machine overhead for the fast path. In other words, if not transaction was started and we're in multiplexing, we don't pay the price of the compiler-generate async state machine (which does show up on the TechEmpower benchmarks).

So handling this case should simply go inside CloseAsync below - the fast path already happens only when we have ConnectorBindingScope.None, so we already skip the fast path if the transaction is undisposed.

Reply:
It does make sense.